Matrix norm for order 1 has been added.

mamakancha opened the pull request:
Fixes #12565

(Follow up of the PR #12567 )

mamakancha committed:
Matrix norm for order 1 has been added.

mamakancha committed:
Space in line 4333 fixed.

mamakancha commented:
@jksuom I worked on your advice. I have used numpy's functions. The code has reduced to mere two lines . Local checks have passed also.

jksuom commented:
Numpy functions should be avoided because not every user has them installed. I would expect that standard Python functions could be used.

mamakancha committed:
Numpy dependent functions removed

Buil-in functions of Python have been used now since numpy might not be available to the user.

mamakancha commented:
@jksuom Yes I forgot to keep that in mind. I have edited the code which now used only Python/Sympy functions. Please have a look at it.

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -4371,8 +4369,13 @@ def norm(self, ord=None):
+                column = [] # keeps track of the number of columns
+
+                for i in range(self.shape[1]) :
+                    column.append(i)
```
Will `column` be a copy of `range(self.cols)`?

mamakancha replied:
Oh Yes. How silly of me not to notice that. Making a list from a list. Redundant steps. I will fix it.

mamakancha committed:
Redundant code removed.

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -4366,7 +4368,11 @@ def norm(self, ord=None):
         else:
-            if ord == 2:  # Spectral Norm
+            if ord == 1:  # Maximum column sum
+                def maxcol(x) : return sum(self.applyfunc(lambda y : abs(y)).col(x))
```
`applyfunc` is run repeatedly for each call of `maxcol`. That is not efficient. I would first save a copy of the matrix with absolute values of entries: 

    m = self.applyfunc(abs).

(`lambda y: abs(y)` is the same function as `abs`.)

Moreover, it is not necessary to introduce a new function `maxcol`.  The list of column sums can be obtained directly from the Python generator expression

    sum(m.col(i)) for i in range(m.cols))

There only remains to apply `max` to this result.

ddjr commented on sympy/matrices/matrices.py:
```diff
@@ -4366,7 +4368,11 @@ def norm(self, ord=None):
-            if ord == 2:  # Spectral Norm
+            if ord == 1:  # Maximum column sum
+                def maxcol(x) : return sum(self.applyfunc(lambda y : abs(y)).col(x))
+                return max(list(map(maxcol, range(self.cols))))
```
`max() ` can directly take generators and therefore `map()` as a parameter. You can simplify your code to

```
return max(map(maxcol, range(self.cols)))
```

gxyd committed:
Fix norm of 'ord=1', add test. Fixes #12565

gxyd commented:
@jksuom please review

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -4366,7 +4368,11 @@ def norm(self, ord=None):
         else:
-            if ord == 2:  # Spectral Norm
+            if ord == 1:  # Maximum column sum
+                return Max(*[sum(self.col(i).applyfunc(abs))
```
I would prefer having `applyfunc` called only once (for the whole matrix) instead of separately for each column.

gxyd committed:
Use 'applyfunc' method once, instead of using it on column submatrices

This is more efficient

jksuom commented:
Thanks, this looks fine to me.

gxyd commented:
Thanks.
